### ISSUE TITLE
FileNotFoundError containing fugitive buffer name

### ISSUE THREAD

**Issue by blueyed** [davidhalter/jedi#1062]
Sometimes Jedi tries to open a file that is not a real file, but looks like it is the cwd plus the buffer name.

It looks like this happens when a fugitive buffer was indexed/cached?! (e.g. when using :Gdiff etc?!)

```
Traceback (most recent call last):
  File "/home/user/.vim/plugged/jedi-vim/pythonx/jedi_vim.py", line 131, in wrapper
    return func(*args, **kwargs)
  File "/home/user/.vim/plugged/jedi-vim/pythonx/jedi_vim.py", line 242, in goto
    definitions = [x for x in script.goto_definitions()
  File "/home/user/.vim/plugged/jedi-vim/pythonx/jedi/jedi/api/__init__.py", line 187, in goto_definitions
    definitions = helpers.evaluate_goto_definition(self._evaluator, context, leaf)
  File "/home/user/.vim/plugged/jedi-vim/pythonx/jedi/jedi/api/helpers.py", line 188, in evaluate_goto_definition
    return evaluator.goto_definitions(context, leaf)
  File "/home/user/.vim/plugged/jedi-vim/pythonx/jedi/jedi/evaluate/__init__.py", line 233, in goto_definitions
    return helpers.evaluate_call_of_leaf(context, name)
  File "/home/user/.vim/plugged/jedi-vim/pythonx/jedi/jedi/evaluate/helpers.py", line 99, in evaluate_call_of_leaf
    values = context.eval_node(base)
  File "/home/user/.vim/plugged/jedi-vim/pythonx/jedi/jedi/evaluate/base_context.py", line 131, in eval_node
    return self.evaluator.eval_element(self, node)
  File "/home/user/.vim/plugged/jedi-vim/pythonx/jedi/jedi/evaluate/__init__.py", line 193, in eval_element
    return self._eval_element_if_evaluated(context, element)
  File "/home/user/.vim/plugged/jedi-vim/pythonx/jedi/jedi/evaluate/__init__.py", line 205, in _eval_element_if_evaluated
    return self._eval_element_cached(context, element)
  File "/home/user/.vim/plugged/jedi-vim/pythonx/jedi/jedi/evaluate/cache.py", line 40, in wrapper
    rv = function(obj, *args, **kwargs)
  File "/home/user/.vim/plugged/jedi-vim/pythonx/jedi/jedi/evaluate/__init__.py", line 209, in _eval_element_cached
    return eval_node(context, element)
  File "/home/user/.vim/plugged/jedi-vim/pythonx/jedi/jedi/debug.py", line 79, in wrapper
    return func(*args, **kwargs)
  File "/home/user/.vim/plugged/jedi-vim/pythonx/jedi/jedi/evaluate/syntax_tree.py", line 48, in wrapper
    return func(context, *args, **kwargs)
  File "/home/user/.vim/plugged/jedi-vim/pythonx/jedi/jedi/evaluate/syntax_tree.py", line 72, in eval_node
    return eval_atom(context, element)
  File "/home/user/.vim/plugged/jedi-vim/pythonx/jedi/jedi/evaluate/syntax_tree.py", line 206, in eval_atom
    search_global=True
  File "/home/user/.vim/plugged/jedi-vim/pythonx/jedi/jedi/_compatibility.py", line 266, in <lambda>
    return lambda *args, **kwargs: self.func(obj, *args, **kwargs)
  File "/home/user/.vim/plugged/jedi-vim/pythonx/jedi/jedi/evaluate/base_context.py", line 148, in py__getattribute__
    return f.find(filters, attribute_lookup=not search_global)
  File "/home/user/.vim/plugged/jedi-vim/pythonx/jedi/jedi/debug.py", line 79, in wrapper
    return func(*args, **kwargs)
  File "/home/user/.vim/plugged/jedi-vim/pythonx/jedi/jedi/evaluate/finder.py", line 67, in find
    types = self._names_to_types(names, attribute_lookup)
  File "/home/user/.vim/plugged/jedi-vim/pythonx/jedi/jedi/evaluate/finder.py", line 178, in _names_to_types
    contexts = ContextSet.from_sets(name.infer() for name in names)
  File "/home/user/.vim/plugged/jedi-vim/pythonx/jedi/jedi/common/context.py", line 34, in from_sets
    sets = list(sets)
  File "/home/user/.vim/plugged/jedi-vim/pythonx/jedi/jedi/evaluate/finder.py", line 178, in <genexpr>
    contexts = ContextSet.from_sets(name.infer() for name in names)
  File "/home/user/.vim/plugged/jedi-vim/pythonx/jedi/jedi/evaluate/filters.py", line 142, in infer
    return self.get_param().infer()
  File "/home/user/.vim/plugged/jedi-vim/pythonx/jedi/jedi/evaluate/filters.py", line 145, in get_param
    params = self.parent_context.get_params()
  File "/home/user/.vim/plugged/jedi-vim/pythonx/jedi/jedi/evaluate/cache.py", line 40, in wrapper
    rv = function(obj, *args, **kwargs)
  File "/home/user/.vim/plugged/jedi-vim/pythonx/jedi/jedi/evaluate/context/function.py", line 244, in get_params
    return self.var_args.get_params(self)
  File "/home/user/.vim/plugged/jedi-vim/pythonx/jedi/jedi/evaluate/arguments.py", line 82, in get_params
    execution_context.tree_node
  File "/home/user/.vim/plugged/jedi-vim/pythonx/jedi/jedi/debug.py", line 79, in wrapper
    return func(*args, **kwargs)
  File "/home/user/.vim/plugged/jedi-vim/pythonx/jedi/jedi/evaluate/dynamic.py", line 83, in search_params
    funcdef
  File "/home/user/.vim/plugged/jedi-vim/pythonx/jedi/jedi/evaluate/cache.py", line 40, in wrapper
    rv = function(obj, *args, **kwargs)
  File "/home/user/.vim/plugged/jedi-vim/pythonx/jedi/jedi/evaluate/utils.py", line 20, in wrapper
    return list(func(*args, **kwargs))
  File "/home/user/.vim/plugged/jedi-vim/pythonx/jedi/jedi/evaluate/dynamic.py", line 118, in _search_function_executions
    evaluator, [module_context], func_string_name):
  File "/home/user/.vim/plugged/jedi-vim/pythonx/jedi/jedi/evaluate/imports.py", line 562, in get_modules_containing_name
    m = check_fs(p)
  File "/home/user/.vim/plugged/jedi-vim/pythonx/jedi/jedi/evaluate/imports.py", line 529, in check_fs
    with open(path, 'rb') as f:
FileNotFoundError: [Errno 2] No such file or directory: '/home/user/projects/foo/fugitive:/home/user/projects/foo/.git/0/foo/conftest.py'

```

The error happens when using jedi-vim's goto, but being in the foo/conftest.py file itself, not the fugitive buffer.
It seems to go away then on some condition.

Jedi version 144a1de.

**Comment by davidhalter** [comment]
That sounds extremely strange. I've never had an issue like that. I cannot explain that at all. Jedi basically works with absolute paths and that's it. Maybe it's some weird interaction with a git plugin that chdirs?

**Comment by blueyed** [comment]
Are you using fugitive yourself?
It creates buffers with non-readable buffers like that - and jedi-vim might pick them up somehow then.

**Comment by blueyed** [comment]
Is https://github.com/davidhalter/jedi/commit/90a226f898702ba7e1cebff425d2d9f2aa0f1eb2 relevant really?
I do not have steps to reproduce this, but could investigate/update then.

**Comment by davidhalter** [comment]
Ok great. I think I understand it now. There's a line in jedi-vim like this:

```
    jedi.settings.additional_dynamic_modules = \
        [b.name for b in vim.buffers if b.name is not None and b.name.endswith('.py')]
```

This is probably the problem. Jedi then continues and tries to open these files. We might need to check in jedi-vim if these files really exist. Longer term I want to get away from this API anyway in jedi towards an API where you can just manage a project.

So IMO a jedi-vim bug. I can also reproduce this if I open a buffer bar.py that doesn't exist on the file system. I just need to complete here in foo.py:

```
def x(param): param. # <--- complete here
```

**Comment by davidhalter** [comment]
Oddly enough, you actually fixed this bug in aa6857d22d59e88a667152c196410df95454326e. Thanks! :)
